src/gather_classes.rs

```diff
@@ -0,0 +1,417 @@
+                    debug!("Failed to extract tuple with error: {err}");
+                }
+            }
+        }
+    }
+    assert!(report.blocks.insert(block_num));
+    Ok(updates)
+}
+
+// Converts recursive calls to value.get("calls") to a boxed iterator.
+fn get_calls<'a>(obj: &'a Value) -> Box<dyn Iterator<Item = &Value> + 'a> {
+    // TODO (#72) Put debug logging here under a core::option_env flag so it is normally hidden.
+    match obj {
+        Value::String(_same) => Box::new(::std::iter::empty()),
+        Value::Array(call_list) => {
+            let nested_calls = call_list
+                .iter()
+                .flat_map(|inner_call| get_calls(inner_call));
+            Box::new(call_list.iter().chain(nested_calls))
+        }
+        Value::Object(obj_map) => {
+            if let Some(Value::Object(diff_map)) = obj_map.get(DIFFERENT) {
+                let base_list = diff_map.get("base").expect("difference should have base");
+                let native_list = diff_map
+                    .get("native")
+                    .expect("difference should have native");
+                // TODO (#72) Compare and get diffs. For now we just chain.
+                Box::new(get_calls(base_list).chain(get_calls(native_list)))
+            } else if let Some(calls_value) = obj_map.get("calls") {
+                get_calls(calls_value)
+            } else if let Some(_val) = obj_map.get("revert_reason") {
+                Box::new(::std::iter::empty())
+            } else {
+                // TODO (#72) Handle errors gracefully.
+                warn!("unexpected: `{}`", obj);
+                panic!();
+            }
+        }
+        // TODO (#72) Handle errors gracefully.
+        _ => panic!("unexpected!"),
+    }
+}
+
+/// returns `Result<(EntryPoint, ClassHash), anyhow::Error>`
+fn get_tuple_from_call(call: &Value) -> Result<(EntryPoint, ClassHash), anyhow::Error> {
+    let mut values = Vec::<FieldElement>::new();
+    for key in ["entry_point_selector", "class_hash"] {
+        let next = call
+            .get(key)
+            .context(format!("call: `{call}`\n-------Extracting key {key}"))?;
+        let felt_version = serde_json::from_value(next.clone())
+            .context(format!("convert value `{}` to felt", next))?;
+        values.push(felt_version);
+    }
+    Ok((values[0], values[1]))
+}
+
+/// Checks that the report corresponds to the block_num.
+/// Then, if no report exists or `overwrite` is `true`, tries to serialize the report.
+fn try_serialize_block_report(
+    block_num: u64,
+    report: &ClassHashesReport,
+    overwrite: bool,
+) -> Result<(), anyhow::Error> {
+    assert!(report.blocks.len() == 1 && report.blocks.contains(&block_num));
+
+    let report_path = path_for_report(block_num);
+    if overwrite || !report_path.exists() {
+        try_serialize(path_for_report(block_num), report).context(format!("block_num: {block_num}"))
+    } else {
+        Err(anyhow!("Skipped"))
+    }
+}
```

Comment:
With the new changes, I believe this function is unnecesary

Reply:
Removed.